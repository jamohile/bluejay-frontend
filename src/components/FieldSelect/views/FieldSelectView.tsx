import * as React from 'react';
import './fieldSelectView.css';
import { string } from 'prop-types';

// Properties for the whole FieldSelectView. This is the highest level component here.
export interface FieldSelectViewProps {
    // Handles a new field being selected.
    onChange: (field:string) => void,

    // The possible fields that can be selected.
    fieldOptions: string[],

    // The currently selected fields.
    selected: string,

    // Whether the field options are being loaded from server.
    loading: boolean,

    // Whether the field options are loaded from server.
    loaded: boolean,

    // Whether something went wrong.
    error:boolean
}

// Properties for the select input inside the FieldSelectView.
export interface OptionViewProps {
    // The selected field
    selected: string,
    // The possible fields
    options: string[],
    // Handle a changed field.
    onChange: (field:string) => void;
}

export class FieldSelectView extends React.Component<FieldSelectViewProps> {
    constructor(props:FieldSelectViewProps){
        super(props);
    }

    render() {

        // Retains type information.
        const {loaded, loading, error, fieldOptions, selected, onChange} = this.props;

        return (
            <div className = {'fieldSelect' + (error ? ' error' : '')}>
                {
                    error &&
                    <ErrorView/>
                }
                {
                    loading &&
                    <LoadingView/>
                }

                {
                    loaded &&
                        <OptionView
                            selected = {selected}
                            options = {fieldOptions}
                            onChange = {onChange}
                        />
                }
            </div>
        )
    }
}

// Lets user know app is loading.
export const LoadingView = () => (
    <span className='faded text'>
        One moment..
    </span>
)

// As select view to choose the field being filtered on.
export const OptionView = ({selected, options, onChange}:OptionViewProps) => (
    <span className = 'optionView'>
        <span className = 'faded text'>
            Filter by
        </span>
        <select className='bold text' onChange={(e:any)=>onChange(e.target.value)}>
            {
                //Iterate through fields, add to the select.
                options.map((op:string) =>  (
                    <option
                        key={op}
                        selected = {selected == op}>
                        {op}
                    </option>
                ))
                }
        </select>
    </span>
)

// Lets user know if something went wrong.
export const ErrorView = () => (
    <span className='text'>
        Something went wrong...
    </span>
)
