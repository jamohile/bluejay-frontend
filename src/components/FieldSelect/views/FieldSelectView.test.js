import React from 'react';
import renderer from 'react-test-renderer';
import {LoadingView, ErrorView, OptionView, FieldSelectView} from './FieldSelectView';

const sampleField = 'someField';
const sampleFieldOptions = ['someField, someOther, Another'];

it('Renders LoadingView.', () => {
    const tree = renderer.create(
        <LoadingView/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('Renders ErrorView.', () => {
    const tree = renderer.create(
        <ErrorView/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('Renders OptionsView.', () => {
    const tree = renderer.create(
        <OptionView
            selected = {sampleField}
            options = {sampleFieldOptions}
            onChange = {(field) => {}}
        />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('Renders loaded FieldSelectView', () => {
    const tree = renderer.create(
        <FieldSelectView
            fieldOptions = {sampleFieldOptions}
            selected = {sampleField}
            onChange = {(field) => {}}
            loaded
        />
    ).toJSON();
    expect(tree).toMatchSnapshot();
})

it('Renders loading FieldSelectView', () => {
    const tree = renderer.create(
        <FieldSelectView
            fieldOptions = {sampleFieldOptions}
            selected = {sampleField}
            onChange = {(field) => {}}
            loading
        />
    ).toJSON();
    expect(tree).toMatchSnapshot();
})

it('Renders error FieldSelectView', () => {
    const tree = renderer.create(
        <FieldSelectView
            fieldOptions = {sampleFieldOptions}
            selected = {sampleField}
            onChange = {(field) => {}}
            error
        />
    ).toJSON();
    expect(tree).toMatchSnapshot();
})

