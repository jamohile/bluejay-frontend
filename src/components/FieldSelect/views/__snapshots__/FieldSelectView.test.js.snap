// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ErrorView. 1`] = `
<span
  className="text"
>
  Something went wrong...
</span>
`;

exports[`Renders LoadingView. 1`] = `
<span
  className="faded text"
>
  One moment..
</span>
`;

exports[`Renders OptionsView. 1`] = `
<span
  className="optionView"
>
  <span
    className="faded text"
  >
    Filter by
  </span>
  <select
    className="bold text"
    onChange={[Function]}
  >
    <option
      selected={false}
    >
      someField, someOther, Another
    </option>
  </select>
</span>
`;

exports[`Renders error FieldSelectView 1`] = `
<div
  className="fieldSelect error"
>
  <span
    className="text"
  >
    Something went wrong...
  </span>
</div>
`;

exports[`Renders loaded FieldSelectView 1`] = `
<div
  className="fieldSelect"
>
  <span
    className="optionView"
  >
    <span
      className="faded text"
    >
      Filter by
    </span>
    <select
      className="bold text"
      onChange={[Function]}
    >
      <option
        selected={false}
      >
        someField, someOther, Another
      </option>
    </select>
  </span>
</div>
`;

exports[`Renders loading FieldSelectView 1`] = `
<div
  className="fieldSelect"
>
  <span
    className="faded text"
  >
    One moment..
  </span>
</div>
`;
